### PR TITLE
Admin team editing

### PR DESCRIPTION
**Pull request by RedbackThomson**
Allows for editing teams through the `Teams` tab of the event page.
- Adds an edit button within each Team card
- Creates a new `PATCH` endpoint for teams that accepts `addMembers` and `removeMembers` lists

### CODE REVIEW

File: src/client/pages/EventPage/tabs/TeamsTab.tsx
Comment (subhankar-panda) [inline src/client/pages/EventPage/tabs/TeamsTab.tsx:269]: Please remove

File: src/server/api/controllers/admin/EventsController.ts
Comment (subhankar-panda) [inline src/server/api/controllers/admin/EventsController.ts:188]: Should we not be removing before adding? I haven't run this myself but what happens if we have users A B C D and then edit to have users A X Y Z. Does this mean that we will make the team have 7 users before we remove B C D? sounds bad.

Comment (subhankar-panda) [reply src/server/api/controllers/admin/EventsController.ts:188]: We should probably be running these 2 operations in a transaction too

**Review by subhankar-panda — COMMENTED** [review]
Haven't looked at FE yet, but I had some concerns w/ backend - lmk

File: src/server/api/controllers/admin/EventsController.ts
Comment (RedbackThomson) [reply src/server/api/controllers/admin/EventsController.ts:188]: Yeah in order to not hit the member limit we would have to remove before adding. This is a soft limit that we impose, but the `User` model requires a Team ID which can only be one value or another.
Transactions were added in Mongoose 5.2.0, so I'll try to add this in, too.

File: src/client/components/EditTeamModal.tsx
Comment (RedbackThomson) [inline src/client/components/EditTeamModal.tsx:82]: @subhankar-panda @nickthegroot  Type unsafe. Would be better to have arrays of `addMembers` and `removeMembers` storing relevant emails, rather than attempting to add/remove `TESCUser` elements to the members array.

File: src/server/api/controllers/admin/EventsController.ts
Comment (RedbackThomson) [inline src/server/api/controllers/admin/EventsController.ts:45]: @subhankar-panda Oooo conditional dependency injection? Yes please.

File: package.json
Comment (RedbackThomson) [inline package.json:94]: I wanted hooks

File: src/client/components/EditTeamModal.tsx
Comment (subhankar-panda) [reply src/client/components/EditTeamModal.tsx:82]: isnt mutation props just plain wrong too?

**Comment by subhankar-panda** [comment]
image
would be nice to have some better errors here but it can come later :) cc @lisalluo @efeichen

File: src/server/utils/Errors.ts
Comment (subhankar-panda) [inline src/server/utils/Errors.ts:34]: Can you make this part of `a` team? Assume we have 2 teams with one user each and i try to add one user to the other team, this text implies that user 2 is a part of team 1 while what we really mean is that they are associated with _another_ team. 

image
image

**Review by subhankar-panda — CHANGES_REQUESTED** [review]
There's a bug when you create a team with one user and remove them. The backend handles it well (team with no users) but the UI does not have the if checks in the teams rendering code to bail out. (It tries to render the user circle which breaks when there are no teammates

File: src/server/api/controllers/admin/EventsController.ts
Comment (lisalluo) [inline src/server/api/controllers/admin/EventsController.ts:184]: could we specify which emails aren't attached to an account in the error message?

File: src/server/utils/Errors.ts
Comment (lisalluo) [reply src/server/utils/Errors.ts:34]: +1 "a team" or "another team" would be clearer

File: src/server/utils/Errors.ts
Comment (lisalluo) [inline src/server/utils/Errors.ts:35]: is USER_NOT_ON_TEAM used anywhere? is it something for the future where we can check to make sure that all users are on a team?

**Review by lisalluo — CHANGES_REQUESTED** [review]
sorry this took me so long to review! thank you 1000000x for doing this :) 

and just to make sure that i didn't misunderstand anything: 

- currently, it's up to the admin/developer/organizer/whatever to make sure that everybody is on a team? 
- and what happens if we need to remove someone from a team and create an entirely new team to put this person on? is there a way to do that currently?

File: src/server/api/controllers/admin/EventsController.ts
Comment (lisalluo) [reply src/server/api/controllers/admin/EventsController.ts:184]: oops i missed panda's comment about this. this can be changed in the future :)

**Comment by lisalluo** [comment]
oops also friendly reminder to please not merge until i confirm that the backups are working! im working on that this week

File: src/server/api/controllers/admin/EventsController.ts
Comment (RedbackThomson) [reply src/server/api/controllers/admin/EventsController.ts:45]: um no, because we can't rely on dependency injections for this - as we conditionally want to choose the service.

File: src/server/utils/Errors.ts
Comment (RedbackThomson) [reply src/server/utils/Errors.ts:35]: Hmm no actually, this string isn't being used anywhere

File: src/server/api/controllers/admin/EventsController.ts
Comment (subhankar-panda) [reply src/server/api/controllers/admin/EventsController.ts:45]: Oh I didn't realize those where args to the constructor! so weird that you can annotate those as private lol

**Comment by RedbackThomson** [comment]
> * currently, it's up to the admin/developer/organizer/whatever to make sure that everybody is on a team?

Yes this does seem like a bit of a flaw. Maybe a UI for people who don't have teams - where you can create new teams - would be great? But at the moment, people could slip through the cracks.

> * and what happens if we need to remove someone from a team and create an entirely new team to put this person on? is there a way to do that currently?

Nope ;(

File: src/server/api/controllers/admin/EventsController.ts
Comment (RedbackThomson) [reply src/server/api/controllers/admin/EventsController.ts:45]: `private` in this case is syntactic sugar for "Create a private member variable and assign it from this parameter".

```
private TransactionService transactionService;

constructor(TransactionService _transactionService...){
   this.transactionService = _transactionService;
}
```

**Comment by lisalluo** [comment]
tested out deleting a team until it's empty and kept getting an error like this:
image

I'm guessing it's trying to load the team after the team is fully deleted?

**Comment by lisalluo** [comment]
> oops also friendly reminder to please not merge until i confirm that the backups are working! im working on that this week

@RedbackThomson @subhankar-panda we're good on this btw! tried restoring to my test db and it looks good :) but to be safe, i'll need to make a backup right before whenever we merge

**Pull request closed without merging**
